### ISSUE TITLE
Document that calling disableOnClickOutside() in componentDidMount will not work

### ISSUE THREAD

**Issue by samsch** [Pomax/react-onclickoutside#71]
As per the title, calling this.props.disableOnClickOutside() from componentDidMount() doesn't disable listening. This is using ES2015 class syntax (but not ES7/decorator).

Here is the relevent code snippet:
```
componentDidMount() {
    this.props.disableOnClickOutside();
}
```
With this code in my component, handleClickOutside is still being called.

[Here is a test case file.](https://gist.github.com/samsch/018bf78fd64d92396f5bb5f6c8583677)
It needs to be compiled with Babel+Webpack. If you want my exact config, you could drop the gist code into my [basic-react repo](https://github.com/samsch/basic-react) as `src/main.js` (the html grabs main.js, not main.jsx, but I used .jsx so that gist would highlight correctly). You would just need to do `npm install react-onclickoutside --save`, and `npm run dev`, then serve `public/` however.

~~I'm looking for a workaround as well. If I find one, I'll report back.~~
You can work around this by putting disableOnClickOutside in a setTimeout.
```
componentDidMount() {
    setTimeout(()=>this.props.disableOnClickOutside(), 0);
}
```

**Comment by Pomax** [comment]
Creating an enabled component, but then immediately disabling it in the `componentDidMount` life cycle functions means you really should have built the component with the disable flag turned on instead. Using the code shown in the the [README.md](https://github.com/Pomax/react-onclickoutside#regulate-whether-or-not-to-listen-for-outside-clicks):

```
class App extends React.Component {
  ...
  render() {
    return ...
      <ABox clicks={this.state.numClicks} handler={this.handleClick} disableOnClickOutside={true} />
     ...;
  }
  ...
```

**Comment by samsch** [comment]
That makes sense practically, but in a real life situation, I don't want to have to set an arbitrary property from my parent. This parent doesn't care if the child is listening or not.

I suppose this would lead to a solution of wrapping the HOCed component in another HOC which sets that property. This seems hacky. Am I missing another obvious solution here?

**Comment by Pomax** [comment]
it's not an arbitrary property, though; the React philosophy is that parent control the bootstrapping of their children by specifying properties, and children control their state (potentially further bootstrapped on the `props` they received).

The HOC should do a straight up pass through of the properties you pass in the "normal" way (did a quick check, the [code suggests that's still true](https://github.com/Pomax/react-onclickoutside/blob/master/decorator.js#L24-L28)), so setting `disableOnClickOutside` should work regardless of whether you're setting it on the true component, or on the HOC-wrapped component. Both should do the right thing.

**Comment by rafeememon** [comment]
@samsch, if you want the child to have complete control over whether to handle outside-click events, can you add logic to your `handleClickOutside` handler to fail fast when appropriate?

**Comment by samsch** [comment]
For now, the setTimeout workaround will work for me. I think what would be nice to close the issue is a small note in the docs that you can't directly call `disableOnClickOutside()` in `componentDidMount` (and have it work). That's what initially tripped me up.

**Comment by Pomax** [comment]
morphing to docs issue (if you have some text in mind let me know and I can add it in, or you can actually hit the edit link for the README, drop the text in, and then tell github to just automagically create a PR for that change. Very nice if you don't want to clone/sync remote/add/commit/push, *then* form a PR!)

**Comment by samsch** [comment]
Ok, so I took a few minutes to test the mixin version. It seems that for createClass components with the mixin, you *can* call `this.disableOnClickOutside()` in `componentDidMount`.

I'm not sure if that's a big enough inconsistency between the APIs to worry about fixing, or just something to toss in the docs.

**Comment by Pomax** [comment]
I'm 100% for "documenting the fact that you should not be creating the component with listening state X, and then upon mount change it to !X", because that's just plain old bad code. Create the element with initial prop !X, done. If you think that the README.md can do with a docs improvement to make that clear, happy to merge documentation improvements, but code fixes will be the absolute worst option here =)

**Comment by samsch** [comment]
I realized that a "duh" solution for the mixin is to set `disableOnClickOutside` to true as a default property on the component. I think this fits closer to the intended pattern?

For the HOC, this approach doesn't work, at least partially due to issue #65.

**Comment by Pomax** [comment]
Sorry for the week long silence, I've been kept busy.... but yes, that's the solution I gave in https://github.com/Pomax/react-onclickoutside/issues/71#issuecomment-203085229 =)

Good point about the HOC having issues here, reading up on those issues now.

**Comment by samsch** [comment]
I think this issue can be closed since pull #76 is merged, or pending #77 as a fix for HOCs (#65).

**Comment by Pomax** [comment]
closing, we can continue the last bits in the PR
